File: Stack/1249. Minimum Remove to Make Valid Parentheses/1249.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        stack = []
        s_list = list(s)

        for i in range(len(s_list)):
            if s_list[i] == '(':
                stack.append(i)
            elif s_list[i] == ')':
                if stack:
                    stack.pop()
                else:
                    s_list[i] = ''
        
        while stack:
            s_list[stack.pop()] = ''

        return ''.join(s_list)

File: Stack/1249. Minimum Remove to Make Valid Parentheses/1249.cpp
#pragma GCC optimize('0fast', 'unroll-loops')

class Solution {
public:
    string minRemoveToMakeValid(string s) {
        ios_base::sync_with_stdio(0);
        cout.tie(0);
        cin.tie(0);

        stack<int> stk;
        string res="";

        for(int i=0; i<s.size(); ++i){
            if(s[i]=='('){
                stk.push(i);
            } else if(s[i]==')'){
                if(!stk.empty()){
                    stk.pop();
                } else{
                    s[i]='&';
                }
            }
        }
        while(!stk.empty()){
            s[stk.top()]='&';
            stk.pop();
        }
        for(char c : s){
            if(c!='&'){
                res+=c;
            }
        }
        return res;
    }
};